Add a 64bit version of the millis() function

jgstroud opened the pull request:
millis() will rollover after 50 days leading to some interesting bugs. Create a 64bit version.

jgstroud commented:
Fixes #261

dkerr64 commented:
Looks like the commit removed code related to syslog port.  Was that intentional?

jgstroud commented:
Nope.  Something must not have merged properly.  I'll rebase and update.

jgstroud commented:
@dkerr64 good catch.  fixed now.  that was strange.  its like part of a single commit was missing from my branch.  rebased again and now I think it looks right.
I'm going to go ahead and keep running with this until I verify the rollover works before I merge it.

mitchjs commented:
out of curiosity, what "interesting bugs" did a 59day rollover create?
as rollover of millis() is not usually problem (as long as everything is unsigned longs and subtraction used)

thanks
Mitch

jgstroud commented:
In particular I saw my device go into recovery mode when I clicked the wall button.  We have a timeout value that should have reset but the millis rolled and the timeout ended up being way off in the future

dkerr64 commented:
> @dkerr64 good catch. fixed now. that was strange. its like part of a single commit was missing from my branch. rebased again and now I think it looks right. I'm going to go ahead and keep running with this until I verify the rollover works before I merge it.

Thanks, looks good now.

dkerr64 commented:
@jgstroud I added this for the ratgdo32 version, you can see the commit [here](https://github.com/ratgdo/homekit-ratgdo32/pull/26/commits/9b65bf2c05470c622132aa7f5683c8483b340ec1).

There were quite a few `unsigned long` variables that I found needed to be change to long long  (or `uint64_t`) which are missed in your changes here.  I suggest a grep for `unsigned long` and/or a look at the changes I made for ratgdo32.

jgstroud commented:
I'll take a look but I made sure it compiled without warning.  I would have expected type mismatch warnings.

dkerr64 commented:
> I'll take a look but I made sure it compiled without warning. I would have expected type mismatch warnings.

I googled and apparently it is a _feature_ of the C compilers to accept assignment to different types, and it will truncate if necessary.  You can have it warn with a `-Wconversion` flag.  So I tried that... bad idea, thousands of warnings from just about every source file (not just ours).

jgstroud commented:
Yeah, it's definitely legal in C.  I really thought it would throw a warning if you didnt explicitly cast the data type, but I think I'm wrong there.  I also forgot that a long is 32 bits on this device.  Pushed new code that hopefully catches the rest of the missed variables

dkerr64 commented:
> Yeah, it's definitely legal in C. I really thought it would throw a warning if you didnt explicitly cast the data type, but I think I'm wrong there. I also forgot that a long is 32 bits on this device. Pushed new code that hopefully catches the rest of the missed variables

Yes, I think you got them all now.  Thanks.

dkerr64 commented:
> > Yeah, it's definitely legal in C. I really thought it would throw a warning if you didnt explicitly cast the data type, but I think I'm wrong there. I also forgot that a long is 32 bits on this device. Pushed new code that hopefully catches the rest of the missed variables
> 
> Yes, I think you got them all now. Thanks.

Spoke too soon. I found another one... `lastDoorUpdateAt` in web.cpp needs to be declared 64-bit.

jgstroud commented:
I had 2 crashes after 10 days.  I haven't analyzed the crash dumps yet to see if it is related to this patch, but leading up to that I had a storm of HK log messages asking for the firmware version.  hundreds per second.  so maybe unrelated.

jgstroud commented:
> Spoke too soon. I found another one... `lastDoorUpdateAt` in web.cpp needs to be declared 64-bit.

I thought I got that one 🤔

jgstroud commented:
@dkerr64 pretty sure this change has introduced some instabilities.  It needs some rethinking before merging.

dkerr64 commented:
ok, that is concerning.  I would not have expected a change from storing a value as 32bit to storing at 64bit to make any difference.

jgstroud commented:
There are only a few places where I think the rollover isn't handled properly.  So, I think I'll fix that and keep with the 32bit version.  I think we should only use the 64bit for the uptime.  The ESP32 probably better handles 64bit.  But if you are seeing crashes, I wouldn't rule this out.

jameyers4 commented:
Is there any update to this issue please @jgstroud ?

jgstroud commented:
No.  I haven't had any time to work on it.

jgstroud commented:
Yes. Please open a new PR.  This is the better approach.  I knew this was the better approach to begin with but I thought I could quickly hack in a fix which was clearly a bad decision for all the reasons you mentioned.

dkerr64 commented:
Closing this as PR #276 has better fix.
